### DISCUSSION TITLE
How to build custom plugin without using template code from separate repository?

### DISCUSSION THREAD

**Discussion by joshas** [WerWolv/ImHex#681] [Get Help] [answered]
[ImHex-Cpp-Plugin-Template](https://github.com/WerWolv/ImHex-Cpp-Plugin-Template) is two versions behind, [ImHex-Rust-Plugin-Template](https://github.com/WerWolv/ImHex-Rust-Plugin-Template) even more. I understand that updating plugin templates is low priority for this project, but considering fast development pace, it should be possible to build plugins using main repository code, right? Could someone provide instructions on how to do it?

**Comment by WerWolv** [comment, marked as answer]
Hey
The templates are specifically made to make plugin development easier. If I haven't updated the templates myself yet, all you need to do is clone the repo and update the ImHex submodule in there. Everything else should just work.
The Rust template is much further behind because nobody ever put the work into making a nice Rust <-> C++ interface for it so there's not a ton you can do in Rust anyways. The C++ template however is fine.

If you start developing a plugin directly in the main repository, you will be forced to update the whole thing and do a bunch of copy-pasting every time I update ImHex which probably isn't what you want.

**Reply by joshas** [reply]
Updated ImHex submodule and tried building, but ended up with error:
````
CMake Error at ImHex/lib/external/pattern_language/external/intervaltree/CMakeLists.txt:6 (add_library):
  add_library cannot create target "intervaltree" because another target with
  the same name already exists.  The existing target is an interface library
  created in source directory
  "~/temp/imhex/ImHex-Cpp-Plugin-Template/ImHex/lib/external/intervaltree".
  See documentation for policy CMP0002 for more details.
````
Building plugin without any modifications works just fine, although it results in plugin only compatible with older ImHex version 1.20.0.

**Reply by joshas** [reply]
My bad, I did not update all of ImHex submodules. But it still is stuck with another error:
`CMake Error at ImHex/lib/libimhex/CMakeLists.txt:85 (set_target_properties):
  set_target_properties Can not find target to add properties to: capstone`

**Reply by WerWolv** [reply]
Are you sure you properly cloned the repository? Try deleting your local repo and clone it again using `git clone https://github.com/WerWolv/ImHex-Cpp-Plugin-Template --recurse-submodules`

**Reply by joshas** [reply]
Deleted old repository, cloned new one - now everything works correctly. Thank you very much for your help.
